[Site Request] Comic Trail #2465

Robonau opened the pull request:
resolves #2465
this took me way too long

Robonau committed:
[Site Request] Comic Trail #2465

resolves #2465
this took me way too long

Robonau committed:
Update ComicTrail.mjs

lint

27136773 commented on src/web/mjs/connectors/ComicTrail.mjs:
```diff
@@ -0,0 +1,54 @@
+        super.id = 'comictrail';
+        super.label = 'Comic Trail (コミックトレイル)';
+        this.tags = [ 'manga', 'japanese' ];
+        this.url = 'https://comic-trail.jp/';
```
Remove trailing `/`

Robonau committed:
Update ComicTrail.mjs

trailing /

27136773 commented on src/web/mjs/connectors/ComicTrail.mjs:
```diff
@@ -0,0 +1,54 @@
+        let request = new Request(this.url + manga.id, this.requestOptions);
+        let data = await this.fetchDOM(request, this.queryChapters);
+        return data.map(element => {
+            let part = /\d{4}-\d{1,2}-\d{1,2}/.exec(element.children[1].firstElementChild.firstElementChild.children[1].textContent);
```
May use `test` instead of `exec`

Robonau committed:
Update ComicTrail.mjs

test not exec

Robonau committed:
Update ComicTrail.mjs

lint

27136773 commented on src/web/mjs/connectors/ComicTrail.mjs:
```diff
@@ -0,0 +1,52 @@
+
+        this.queryManga = 'section.section div.columns div.column div > a';
+
+        this.queryChapters = 'section.section > ul.columns > li.column.is-full-mobile.is-half-tablet.chapter-item div.columns.is-mobile.is-gapless.chapter-item__inner';
```
Query can be simplified
`'ul.chapters li.chapter-item div.chapter-item__body'`

Robonau replied:
it can me simplified but that one sure doesn't work

27136773 replied:
I tested it in browser and it worked fine, except it goes one level deeper than the original query and therefore needs also adjustments in the chapter queries, e.g.
`element.child[0].child[3].child[1].textContent` would be `element.child[3].child[1].textContent`

27136773 commented on src/web/mjs/connectors/ComicTrail.mjs:
```diff
@@ -0,0 +1,52 @@
+                part = ` ${element.children[1].firstElementChild.firstElementChild.children[1].textContent.trim()}`;
+            }
+            return {
+                id: this.getRootRelativeOrAbsoluteLink(`${element.firstElementChild.firstElementChild.pathname}/` || "", this.url),
```
Get ID with CSS query
`element.querySelector('a.button.is-read').href`

Robonau replied:
needs avoidance if, if element.querySelector('a.button.is-read') = null 
cause no like null.href

27136773 commented on src/web/mjs/connectors/ComicTrail.mjs:
```diff
@@ -0,0 +1,52 @@
+            }
+            return {
+                id: this.getRootRelativeOrAbsoluteLink(`${element.firstElementChild.firstElementChild.pathname}/` || "", this.url),
+                title: element.children[1].firstElementChild.firstElementChild.firstElementChild.textContent.trim() + part
```
Get title with CSS query
`element.querySelector('p.has-text-weight-bold').textContent.trim()`

27136773 replied:
I think i understand the concern about the part, in that case the query could be like this:
`title: [...element.querySelectorAll('p.has-text-weight-bold')].map(p => p.textContent.trim()).join(' ');`

Grab all parts and concatenate them with a space...

27136773 commented on src/web/mjs/connectors/ComicTrail.mjs:
```diff
@@ -0,0 +1,52 @@
+        let data = await this.fetchDOM(request, this.queryChapters);
+        return data.map(element => {
+            let part = "";
+            if (!(/\d{4}-\d{1,2}-\d{1,2}/.test(element.children[1].firstElementChild.firstElementChild.children[1].textContent))){
```
No sure what the reason behind this is, i would just remove it and get id/title with CSS query.

Robonau replied:
the second line is sometimes a date like "2019-9-11" and sometimes like "part 2", the "part 2" is necessary the date is not

Robonau replied:
so, if (not date) put at end of title

Robonau replied:
yea, you right just grab both with a queryall

27136773 commented on src/web/mjs/connectors/ComicTrail.mjs:
```diff
@@ -0,0 +1,52 @@
+        let data = await this.fetchDOM(request, this.queryManga);
+        return data.map(element => {
+            return {
+                id: this.getRootRelativeOrAbsoluteLink(element.href, request.url),
```
If element is an `anchor` the `.href` is redundant and can be removed

27136773 commented on src/web/mjs/connectors/ComicTrail.mjs:
```diff
@@ -0,0 +1,52 @@
+        return data.map(element => {
+            return {
+                id: this.getRootRelativeOrAbsoluteLink(element.href, request.url),
+                title: element.children[0].attributes[2].value.trim()
```
Isn't there a CSS query to get the title?
`element.querySelector('... input[title]').title.trim()`
If not, get attribute by name
`.getAttribute('title')` instead of index `[2]`

Robonau replied:
doesn't like querySelector

Robonau committed:
Update ComicTrail.mjs

queryChapters simplified
used querySelector where possable
.getAttribute('alt')

27136773 commented on src/web/mjs/connectors/ComicTrail.mjs:
```diff
@@ -39,13 +39,15 @@ export default class ComicTrail extends SpeedBinb {
+                id = `${element.querySelector('a.button.is-read').href}/`;
             }
+            let title = '';
+            element.querySelectorAll('p.has-text-weight-bold').forEach(element => title=title.concat(element.textContent.trim()));
```
Do not use outer variable name `element` as scoped inner variable name
`element.querySelectorAll(...).forEach(el => el ...)`

27136773 replied:
Should there be a space between the concatenated parts?
Maybe use `el => title += ' ' + el.textContent.trim()`
or `title = [...element.querySelectorAll('p.has-text-weight-bold')].join(' ')` instead

27136773 commented on src/web/mjs/connectors/ComicTrail.mjs:
```diff
@@ -39,13 +39,15 @@ export default class ComicTrail extends SpeedBinb {
-            if (!/\d{4}-\d{1,2}-\d{1,2}/.test(element.children[1].firstElementChild.firstElementChild.children[1].textContent)){
-                part = ` ${element.children[1].firstElementChild.firstElementChild.children[1].textContent.trim()}`;
+            let id = "";
+            if (element.querySelector('a.button.is-read') != null){
```
This should never be `null`, so the `if` check is obsolete, otherwise there would be chapters with empty IDs which will throw an exception sooner or later anyway.

Robonau replied:
its there because a lot of chapters don't have IDs right now (u gotta pay for them)
i don't know how to deal with them

Robonau committed:
Update ComicTrail.mjs

Robonau committed:
fixes, sorted no id chapters

Robonau committed:
thank you filter method

27136773 reviewed (approved):
Looks good
